### PR TITLE
Combine profiling start/stop APIs into one

### PR DESCRIPTION
**Pull request by poornas**
Take profile duration as a query parameter for profile API

## Motivation and Context
Better user experience 

## How to test this PR?
mc support profile alias

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [ ] Unit tests added/updated

### CODE REVIEW

**Review by harshavardhana — CHANGES_REQUESTED** [review]
We should keep the existing APIs as is and add new APIs for newer functionality.

This API has been around for years breaking in this manner will break a mc, madmin-go etc

**Comment by poornas** [comment]
> We should keep the existing APIs as is and add new APIs for newer functionality.
> 
> This API has been around for years breaking in this manner will break a mc, madmin-go etc

done

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
{"detail": "Not found"}Deleting image locally

**Review by klauspost — APPROVED** [review]
LGTM

File: cmd/admin-handlers.go [hunk 1]
```diff
@@ -594,6 +595,90 @@ func (a adminAPIHandlers) StartProfilingHandler(w http.ResponseWriter, r *http.R
...
+	if dstr := r.Form.Get("duration"); dstr != "" {
+		var err error
+		duration, err = time.ParseDuration(dstr)
+		if err != nil {
+			writeErrorResponseJSON(ctx, w, errorCodes.ToAPIErr(ErrServerNotInitialized), r.URL)
+			return
+		}
+	}
+	io.Copy(ioutil.Discard, r.Body)
+
```
Comment (harshavardhana) [inline cmd/admin-handlers.go:630]: So if I send an unending stream this API will keep reading until EOF - why is this needed?

File: cmd/admin-handlers.go [hunk 2]
```diff
@@ -594,6 +595,90 @@ func (a adminAPIHandlers) StartProfilingHandler(w http.ResponseWriter, r *http.R
...
+	globalProfilerMu.Lock()
+
+	if globalProfiler == nil {
+		globalProfiler = make(map[string]minioProfiler, 10)
+	}
+
+	// Stop profiler of all types if already running
+	for k, v := range globalProfiler {
+		for _, p := range profiles {
```
Comment (harshavardhana) [inline cmd/admin-handlers.go]: You are only stopping profiles that are currently running vs what was asked to run - you may end up leaving older profile running and not stopping them. 

You should blindly stop all profiles not the ones based on what was asked in the API call

File: cmd/admin-handlers.go [hunk 3]
```diff
@@ -594,6 +595,90 @@ func (a adminAPIHandlers) StartProfilingHandler(w http.ResponseWriter, r *http.R
...
+
+		// Start profiling locally as well.
+		prof, err := startProfiler(profiler)
+		if err == nil {
+			globalProfiler[profiler] = prof
+		}
+	}
+	globalProfilerMu.Unlock()
+
+	if globalNotificationSys == nil {
```
Comment (harshavardhana) [inline cmd/admin-handlers.go]: This check is already present..

File: cmd/admin-handlers.go [hunk 4]
```diff
@@ -594,6 +595,90 @@ func (a adminAPIHandlers) StartProfilingHandler(w http.ResponseWriter, r *http.R
...
+	}
+
+	timer := time.NewTimer(duration)
+	for {
+		select {
+		case <-ctx.Done():
+			if !timer.Stop() {
+				<-timer.C
+			}
+			return
```
Comment (harshavardhana) [inline cmd/admin-handlers.go:666]: If context was canceled we should stop the profile and discard them.

**Review by harshavardhana — CHANGES_REQUESTED** [review]

File: cmd/admin-handlers.go (hunk 1 above)
Comment (poornas) [reply cmd/admin-handlers.go:630]: when the client closes connection, the context will not get canceled unless the body is read on POST requests

Comment (harshavardhana) [reply cmd/admin-handlers.go:630]: But this is not the case when body is empty , we should have no body in this API.

Not specifying an upper limit on the r.body the io.Copy potentially won't ever end in the current PR code

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

File: cmd/admin-handlers.go (hunk 2 above)
Comment (poornas) [reply cmd/admin-handlers.go]: done

**Review by harshavardhana — APPROVED** [review]
